File: src/Single_Number_II.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def singleNumber(self, nums):
        ret = 0
        for i in range(32):
            cnt = 0
            for num in nums:
                cnt += 1 if abs(num) & (1 << i) else 0
            cnt %= 3
            ret += (cnt << i)
        cnt = 0
        for num in nums:
            if num == ret:
                cnt += 1
        return ret if cnt == 1 else -ret


if __name__ == "__main__":
    solution = Solution()
    print solution.singleNumber([1, 1, 1, 2, 3, 2, 2])
    print solution.singleNumber([-2, -2, 1, 1, -3, 1, -3, -3, -4, -2])

File: src/algorithms/cpp/Single_Number_II.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        int ret = 0;
        for (int i = 31; i >= 0; -- i) {
            int cnt = 0, tmp = 1 << i;
            for (auto num: nums) {
                if ((tmp & num) != 0)  ++ cnt;
            }
            if (cnt % 3 != 0)   ret = (ret << 1) + 1;
            else                ret <<= 1;
        }
        return ret;
    }
};

int main() {
    vector <int> input {0, 0, 0, 5};
    Solution solution = Solution();
    cout << solution.singleNumber(input) << endl;
    return 0;
}
